src/hellogo_fares/utils.rs

```diff
@@ -0,0 +1,954 @@
+        }
+    }
+}
+
+pub fn get_value_in_keylist<F>(element: &Element, key: &str) -> Result<F>
+where
+    F: FromStr,
+{
+    let values = element
+        .try_only_child("KeyList")?
+        .children()
+        .filter(|key_value| match key_value.try_only_child("Key") {
+            Ok(k) => k.text() == key,
+            _ => false,
+        })
+        .map(|key_value| key_value.try_only_child("Value"))
+        .collect::<Result<Vec<_>>>()?;
+    if values.len() != 1 {
+        bail!(
+            "Failed to find a unique key '{}' in '{}'",
+            key,
+            element.name()
+        )
+    } else {
+        values[0]
+            .text()
+            .parse()
+            .map_err(|_| format_err!("Failed to get the value out of 'KeyList' for key '{}'", key))
+    }
+}
+
+pub fn get_amount_units_factor(element: &Element) -> Result<Decimal> {
+    let amount = element.try_only_child("Amount")?.text();
+    let amount: Decimal = amount
+        .parse()
+        .map_err(|_| format_err!("Failed to convert '{}' into a 'Decimal'", amount))?;
+    let units = element.try_only_child("Units")?.text();
+    let units: Decimal = units
+        .parse()
+        .map_err(|_| format_err!("Failed to convert '{}' into a 'Decimal'", units))?;
+    Ok(amount * units)
+}
+
+pub fn get_unit_price(unit_price_frame: &Element) -> Result<Decimal> {
+    let geographic_interval_price = unit_price_frame
+        .try_only_child("fareStructures")?
+        .try_only_child("FareStructure")?
+        .try_only_child("geographicalIntervals")?
+        .try_only_child("GeographicalInterval")?
+        .try_only_child("prices")?
+        .try_only_child("GeographicalIntervalPrice")?;
+    Ok(get_amount_units_factor(geographic_interval_price)?)
+}
+
+const DATE_TIME_FORMAT: &str = "%+";
+pub fn get_validity(resource_frame: &Element) -> Result<(Date, Date)> {
+    fn extract_date(element: &Element, date_element_name: &str) -> Result<Date> {
+        let date_str = element.try_only_child(date_element_name)?.text();
+        let date = NaiveDate::parse_from_str(date_str.as_str(), DATE_TIME_FORMAT)
+            .map_err(|_| format_err!("Failed to convert '{}' into a 'Date'", date_str))?;
+        Ok(date)
+    }
+
+    if resource_frame.name() != "ResourceFrame" {
+        bail!(
+            "Failed to get validity from a '{}', it should be a 'ResourceFrame'",
+            resource_frame.name()
+        );
+    }
+    let version = resource_frame
+        .try_only_child("versions")?
+        .try_only_child("Version")?;
+    let validity_start_date = extract_date(version, "StartDate")?;
+    let validity_end_date = extract_date(version, "EndDate")?;
+    Ok((validity_start_date, validity_end_date))
+}
+
+pub fn get_currency(fare_frame: &Element) -> Result<String> {
+    let currency = fare_frame
+        .try_only_child("FrameDefaults")?
+        .try_only_child("DefaultCurrency")?
+        .text();
+    if iso4217::alpha3(currency.as_str()).is_some() {
```

Comment:
or early return, as you want

```rust
if iso4217::alpha3(currency.as_str()).is_none() {
    bail!("Failed to validate '{}' as a currency", currency);
}

Ok(currency)
```

Reply:
Done in 3ccab5336d49fc0bfabe64813654a58ae206a90e.